### PR TITLE
2.x: add doFinally to the rest of the reactive base classes

### PR DESCRIPTION
**Pull request by akarnokd**
This PR adds the `doFinally` operator to the rest of the base reactive classes.

### CODE REVIEW

File: src/main/java/io/reactivex/internal/operators/completable/CompletableDoFinally.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            runFinally();
+        }
+
+        @Override
+        public boolean isDisposed() {
+            return d.isDisposed();
+        }
+
+        void runFinally() {
+            if (compareAndSet(0, 1)) {
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/internal/operators/completable/CompletableDoFinally.java:96]: Is there reasoning behind why int over boolean was used?

**Review by JakeWharton — APPROVED** [review]

File: src/main/java/io/reactivex/internal/operators/completable/CompletableDoFinally.java (hunk 1 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/internal/operators/completable/CompletableDoFinally.java:96]: So the pattern stays the same. Internally, AtomicBoolean does the same thing.

Comment (JakeWharton) [reply src/main/java/io/reactivex/internal/operators/completable/CompletableDoFinally.java:96]: The half serialization pattern?

Comment (akarnokd) [reply src/main/java/io/reactivex/internal/operators/completable/CompletableDoFinally.java:96]: This is an "enter once" pattern.

**Pull request merged**